solls/LanguageServer.cpp

```diff
@@ -0,0 +1,679 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#include <solls/LanguageServer.h>
+
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/interface/ReadFile.h>
+
+#include <liblangutil/SourceReferenceExtractor.h>
+
+#include <libsolutil/Visitor.h>
+#include <libsolutil/JSON.h>
+
+#include <boost/filesystem.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+
+#include <ostream>
+
+#include <iostream>
+#include <string>
+
+using namespace std;
+using namespace std::placeholders;
+
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+
+namespace solidity {
+
+namespace // {{{ helpers
+{
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ASTNodeLocator : public ASTConstVisitor
+{
+private:
+	int m_pos = -1;
+	ASTNode const* m_currentNode = nullptr;
+
+public:
+	explicit ASTNodeLocator(int _pos): m_pos{_pos}
+	{
+	}
+
+	ASTNode const* closestMatch() const noexcept { return m_currentNode; }
+
+	bool visitNode(ASTNode const& _node) override
+	{
+		if (_node.location().start <= m_pos && m_pos <= _node.location().end)
+		{
+			m_currentNode = &_node;
+			return true;
+		}
+		return false;
+	}
+};
+
+// TOOD: maybe use SimpleASTVisitor here, if that would be a simple free-fuunction :)
+class ReferenceCollector: public frontend::ASTConstVisitor
```

Comment:
Is this fast?

Reply:
I mean. I'm only using rather small test files. I think it's a start, and I think if it's becoming a problem we could then create a symbol-to-refs map after compile stage in the LSP to speed this up. the trade-off would be more memory consumption on big sources. I don't know if it's still not worth it.